src/executor/evaluate/mod.rs

```diff
@@ -170,6 +172,100 @@ pub async fn evaluate<'a, T: 'static + Debug>(
+                    Evaluated::Literal(l) => Some(l),
+                    _ => None,
+                })
+                .collect::<Option<Vec<_>>>()
+            {
+                None => {
+                    let when_then_value = when_then
+                        .into_iter()
+                        .map(|(when, then)| -> Result<(Value, Value)> {
+                            match (when.try_into(), then.try_into()) {
+                                (Ok(w), Ok(t)) => Ok((w, t)),
+                                (Err(e), _) | (_, Err(e)) => Err(e),
+                            }
+                        })
+                        .collect::<Result<Vec<(Value, Value)>>>()?;
+
+                    let wt_type = when_then_value
+                        .iter()
+                        .map(|(_, then)| discriminant(then))
+                        .collect_vec();
+
+                    if !wt_type.iter().all_equal() {
+                        return Err(EvaluateError::UnequalResultTypes("CASE".to_owned()).into());
+                    }
+                    let else_result = match else_result {
+                        Some(Evaluated::Value(cv)) => {
+                            let er = cv.into_owned();
+                            if Some(&discriminant(&er)) != wt_type.first() {
+                                return Err(
+                                    EvaluateError::UnequalResultTypes("CASE".to_owned()).into()
+                                );
+                            }
+                            Some(Evaluated::from(er))
+                        }
+                        _ => else_result,
+                    };
+
+                    (
+                        when_then_value
+                            .into_iter()
+                            .map(|(when, then)| (Evaluated::from(when), Evaluated::from(then)))
+                            .collect::<Vec<_>>(),
+                        else_result,
+                    )
+                }
+                Some(wt_type) => {
+                    let wt_type = wt_type.into_iter().map(discriminant).collect_vec();
+                    if !wt_type.iter().all_equal()
+                        || match &else_result {
+                            Some(Evaluated::Literal(er)) => {
+                                Some(&discriminant(er)) != wt_type.first()
+                            }
+                            _ => false,
+                        }
+                    {
+                        return Err(EvaluateError::UnequalResultTypes("CASE".to_owned()).into());
+                    }
+
+                    (when_then, else_result)
+                }
+            };
+
+            Ok(when_then
+                .into_iter()
+                .find_map(|(when, then)| when.eq(&operand).then(|| then))
+                .unwrap_or_else(|| match else_result {
+                    Some(er) => er,
+                    None => Evaluated::from(Value::Null),
+                }))
```

Comment:
`Literal` and `Value` can be mixed in `then`, so
```rust
let (when_then, else_result) = match when_then
    .iter()
    .map(|(_, then)| match then {
        Evaluated::Literal(l) => Some(l),
        _ => None,
    })
    .collect::<Option<Vec<_>>>()
```
this above looks unnecessary.

fyi.
 ```rust
 if !when_then                                 
     .iter()                                                    
     .map(|(_, then)| then.try_into().map(discriminant))        
     .all_equal()                                               
 {                                                              
     return Err(EvaluateError::UnequalResultTypes("CASE".to_owned()).into());
 }                                                          
                                                            
 for (when, then) in when_then.into_iter() {                
     if when == operand {                                   
         return Ok(then);                                   
     }                                                      
 }                                                                           
                                                           
 Ok(else_result.unwrap_or_else(|| Evaluated::from(Value::Null)))
```

This is still not complete code, but it passes the current test `case`.

Reply:
> `Literal` and `Value` can be mixed in `then`, so
> 
> ```rust
> let (when_then, else_result) = match when_then
>     .iter()
>     .map(|(_, then)| match then {
>         Evaluated::Literal(l) => Some(l),
>         _ => None,
>     })
>     .collect::<Option<Vec<_>>>()
> ```
> 
> this above looks unnecessary.

They can be mixed, but then we have to use `try_into()` for every `then` values, no matter the enum variance. When every `then` values are `Evaluated::Literal` type, there's no need to use `try_into()` so I thought it would be better split the conditions as:
1. When all the `then` values are `Evaluated::Literal(l)` -> Get the variant value `l` and type check
2. When there's at least one `Evaluated::Value(v)` -> Squeeze the value out with `try_into()` and type check

Well, I'd be glad if using `try_into()` for every case isn't a bad idea. Because my logic here has produced bunch of ugly condition branched codes :< Please let me know if there's something wrong about my opinion 👍

Reply:
Oh.. ok now I got why you wanted to handle `Literal` separately, thanks for sharing the details.
Then.. let's think about merge these `then` comparing logic into a single statement :)
How about using window function?
Rather than using `all_equal` in here, we can use `try_fold` for the alternative for window.
Or.. simply you can try

Reply:
Thanks for the feedback! I'll try shrinking code a lot 📄